### PR TITLE
startup: support init.lua as user config

### PR DESCRIPTION
**Pull request by dm1try**
ref #7895, continuation of #8720

### CODE REVIEW

**Comment by dm1try** [comment]
some bullet points from #8720 are still valid:
> need to confirm this new code interacts correctly with other parts of the startup sequence like $VIMINIT, $EXINIT, 'exrc', $VIM/sysinit.vim

there is no `sysinit.lua` and `exrc` is ignored if lua config in use, atm

> updating documentation (and probably write a tutorial on how to migrate an init.vim to init.lua)

maybe someone will help ;)
and I'm going to write a script for semi-automated migration

**Review by tjdevries — CHANGES_REQUESTED** [review]
This looks cool -- I can take a look at reviewing if you're interested.

We would need to update the docs though if we're going to ship this.

**Comment by dm1try** [comment]
> I can take a look at reviewing if you're interested.

sure, feel free

> We would need to update the docs though if we're going to ship this.

I'm going to update the docs this week then)

**Comment by tjdevries** [comment]
Cool :) I'll try and take a look sometime this week then.

File: test/unit/path_spec.lua
Comment (dm1try) [inline test/unit/path_spec.lua:612]: this helper function is not needed actually because of using `const char*`(according to http://luajit.org/ext_ffi_semantics.html#convert_fromlua) but I used the example from `path_is_absolute`

**Comment by dm1try** [comment]
> I'm going to update the docs this week then)

I mentioned "init.lua" in starting.txt but there are many references to "init.vim" in other places(not sure how properly to handle this)

**Comment by tjdevries** [comment]
We should probably mention the load order (so, `init.lua` overrides `init.vim`).

I think we could probably just have a section that says something along the lines of "throughout the docs, init.vim is used to describe the initialization config file. If you have an `init.lua`, that will take the place of `init.vim`."

The other option would be to replace all the references of `init.vim` with `|init-vim|` or something... I'm not sure what the right name would be. Do you have any suggestions or thoughts?

**Comment by kamalmarhubi** [comment]
`|init-file|` perhaps? or `|startup-file|`?

**Comment by dm1try** [comment]
> We should probably mention the load order (so, init.lua overrides init.vim).

yeah, something like this "It is an error to use both at the same time, in this case only "init.lua" will be loaded." ?

> I think we could probably just have a section that says something along the lines of "throughout the docs, init.vim is used to describe the initialization config file. If you have an init.lua, that will take the place of init.vim."
The other option would be to replace all the references of init.vim with |init-vim| or something... I'm not sure what the right name would be.

the first one is easier, though the second looks like more appropriate.

> Do you have any suggestions or thoughts?

in addition to @kamalmarhubi 's comment, `|config-file|` / `|user-config-file|`

**Comment by tjdevries** [comment]
Can we bring the discussion for the doc naming thing back to the initial issue thread? It will be easier to get everyone involved there again and I do not feel confident enough to make a decision myself. Is that OK?

**Comment by kamalmarhubi** [comment]
Sounds like consensus is `config`:

@teto in https://github.com/neovim/neovim/issues/7895#issuecomment-670976473
> I just noticed that along with the `init.vim` tag there is the `config` tag. So we could just use that one. Replacing |init.vim| by |config| is also more readable in the case I've seen (e.g. `to your |init.vim| file.`)


@tjdevries in https://github.com/neovim/neovim/issues/7895#issuecomment-671101211
> Works for me. I have no strong preference, just wanted to make sure that we all agreed on it. :smile:

**Comment by bfredl** [comment]
rebased. changed to `nlua_exec_file` name.

**Comment by bfredl** [comment]
@tjdevries what do you think about the documentation? Needs it more changes before merge? (we will for sure change it even more before 0.5).

**Comment by dm1try** [comment]
> rebased. changed to nlua_exec_file name.

thanks!

> @tjdevries what do you think about the documentation? Needs it more changes before merge? (we will for sure change it even more before 0.5).

I'm going to re-visit the proposals related to the docs this weekend(this may be a separate  PR, for sure)

**Comment by dm1try** [comment]
rebased and [updated `starting.txt` with a generic config name](https://github.com/neovim/neovim/pull/12235/commits/133971363a8b4ff41f7353f9b1998f4b742fc0a1) ⏩

**Comment by bfredl** [comment]
LGTM. anything missing here (that cannot wait to follow-up PR:s)?

**Comment by bfredl** [comment]
ping @tjdevries time to wake up

**Comment by fsouza** [comment]
wake up @tjdevries

**Comment by tjdevries** [comment]
I have awaken and I grant my approval.

image

**Comment by glepnir** [comment]
LGTM

**Pull request merged**
